File: windows-cpp/src/sessionpane.cpp
#include "sessionpane.h"
#include "statsworker.h"
#include "terminalwidget.h"
#include "sshsession.h"
#include "remotebrowser.h"

#include <QCheckBox>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QVBoxLayout>

SessionPane::SessionPane(
    const QString &name_,
    const QString &fontFamily,
    int fontSize,
    const QString &cursorStyle,
    QWidget *parent)
    : QWidget(parent)
    , name(name_)
{
    cwdTracker = new CwdTracker(this);

    terminal = new TerminalWidget(80, 24, fontFamily, fontSize, cursorStyle, this);

    // Shown only while Multi-Exec is active — it means nothing otherwise.
    excludeCheckbox = new QCheckBox("Exclude from Multi-Exec", this);
    excludeCheckbox->setVisible(false);

    // Offered only after a session drops.
    reconnectBtn = new QPushButton("Reconnect", this);
    reconnectBtn->setVisible(false);
    connect(reconnectBtn, &QPushButton::clicked, this, &SessionPane::reconnectRequested);

    // No title label and no Close button: the tab already carries the session
    // name and a close affordance, so repeating them inside the pane is just
    // chrome around the terminal.
    m_controls = new QWidget(this);
    QHBoxLayout *ctrlLayout = new QHBoxLayout(m_controls);
    ctrlLayout->setContentsMargins(6, 4, 6, 4);
    ctrlLayout->setSpacing(8);
    ctrlLayout->addWidget(excludeCheckbox);
    ctrlLayout->addStretch();
    ctrlLayout->addWidget(reconnectBtn);
    m_controls->setVisible(false);

    QVBoxLayout *layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(0);
    layout->addWidget(terminal, 1);
    layout->addWidget(m_controls);

    connect(terminal, &TerminalWidget::dataToSend, this, &SessionPane::dataToSend);
    connect(terminal, &TerminalWidget::dataToSend, cwdTracker, &CwdTracker::feedInput);
    connect(terminal, &TerminalWidget::sizeChanged, this, &SessionPane::sizeChanged);
}

SessionPane::~SessionPane() {
    disconnectSession();
}

void SessionPane::applySettings(const QString &fontFamily, int fontSize, const QString &cursorStyle) {
    terminal->applySettings(fontFamily, fontSize, cursorStyle);
}

void SessionPane::setReconnectVisible(bool on) {
    reconnectBtn->setVisible(on);
    updateControlsVisibility();
}

void SessionPane::setMultiExecControlsVisible(bool on) {
    excludeCheckbox->setVisible(on);
    updateControlsVisibility();
}

void SessionPane::updateControlsVisibility() {
    m_controls->setVisible(reconnectBtn->isVisible() || excludeCheckbox->isVisible());
}

void SessionPane::startStatsWorker() {
    if (m_statsWorker) return;
    // Stats are gathered by running commands over the SSH channel, so there is
    // nothing to poll for a local WSL session.
    SSHSession *ssh = qobject_cast<SSHSession*>(session);
    if (!ssh) return;
    m_statsWorker = new RemoteStatsWorker(
        ssh->rawSession(), ssh->sessionLock(), this);
    connect(m_statsWorker, &RemoteStatsWorker::statsReady,
            this, [this](const QJsonObject &stats) {
        lastStats = stats;
        emit statsUpdated(stats);
    });
    m_statsWorker->start();
}

void SessionPane::stopStatsWorker() {
    if (!m_statsWorker) return;
    m_statsWorker->stop();
    if (!m_statsWorker->wait(3000))
        m_statsWorker->terminate();
    m_statsWorker->wait(1000);
    delete m_statsWorker;
    m_statsWorker = nullptr;
    lastStats = {};
}

void SessionPane::disconnectSession() {
    stopStatsWorker();
    if (session) {
        session->stop();
        if (!session->wait(5000)) {
            session->terminate();
            session->wait(2000);
        }
        session->deleteLater();
        session = nullptr;
    }
}

File: windows/star_term/session_pane.py
from PySide6.QtCore import Signal
from PySide6.QtWidgets import (
    QCheckBox, QHBoxLayout, QLabel, QPushButton, QVBoxLayout, QWidget,
)

from .remote_browser import CwdTracker
from .ssh_session import RemoteStatsWorker
from .terminal_widget import TerminalWidget


class SessionPane(QWidget):
    """A single SSH session: terminal, exclude-from-multi-exec checkbox, and
    close button. Owns the SSHSession once connected."""

    data_to_send = Signal(bytes)
    size_changed = Signal(int, int)  # cols, rows
    close_requested = Signal()
    reconnect_requested = Signal()
    stats_updated = Signal(dict)

    def __init__(self, name, font_family="Courier New", font_size=10,
                 cursor_style="underline", parent=None):
        super().__init__(parent)
        self.name = name
        self.session = None
        self.connection_params = None
        self.stats_worker = None
        self.last_stats = None
        self.cwd_tracker = CwdTracker()

        self.title_label = QLabel(name)
        self.title_label.setObjectName("sectionTitle")

        self.terminal = TerminalWidget(
            font_family=font_family, font_size=font_size, cursor_style=cursor_style,
        )

        self.exclude_checkbox = QCheckBox("Exclude from Multi-Exec")

        self.reconnect_btn = QPushButton("Reconnect")
        self.reconnect_btn.setVisible(False)
        self.reconnect_btn.clicked.connect(self.reconnect_requested)

        close_btn = QPushButton("Close")
        close_btn.clicked.connect(self.close_requested)

        controls = QWidget()
        controls_layout = QHBoxLayout(controls)
        controls_layout.setContentsMargins(0, 0, 0, 0)
        controls_layout.setSpacing(8)
        controls_layout.addWidget(self.exclude_checkbox)
        controls_layout.addStretch()
        controls_layout.addWidget(self.reconnect_btn)
        controls_layout.addWidget(close_btn)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(6, 6, 6, 6)
        layout.setSpacing(4)
        layout.addWidget(self.title_label)
        layout.addWidget(self.terminal, 1)
        layout.addWidget(controls)

        self.terminal.data_to_send.connect(self.data_to_send)
        self.terminal.data_to_send.connect(self.cwd_tracker.feed_input)
        self.terminal.size_changed.connect(self.size_changed)

    def apply_settings(self, font_family, font_size, cursor_style):
        self.terminal.apply_settings(font_family, font_size, cursor_style)

    def start_stats_worker(self):
        if self.session and self.session.client and self.stats_worker is None:
            self.stats_worker = RemoteStatsWorker(self.session.client)
            self.stats_worker.stats_ready.connect(self._on_stats_ready)
            self.stats_worker.start()

    def _on_stats_ready(self, stats: dict):
        self.last_stats = stats
        self.stats_updated.emit(stats)

    def _stop_stats_worker(self):
        if self.stats_worker:
            self.stats_worker.stop()
            self.stats_worker.wait()
            self.stats_worker = None

    def disconnect_session(self):
        self._stop_stats_worker()
        self.last_stats = None
        if self.session:
            self.session.stop()
            self.session.wait()
            self.session = None
